lib/src/Uncertainty/Process/DiscreteMarkovChain.cxx

```diff
@@ -0,0 +1,329 @@
+ *
+ *  This library is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Lesser General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Lesser General Public License
+ *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include <iostream>
+#include "openturns/DiscreteMarkovChain.hxx"
+#include "openturns/PersistentObjectFactory.hxx"
+#include "openturns/UserDefined.hxx"
+#include "openturns/Dirac.hxx"
+#include "openturns/ResourceMap.hxx"
+#include "openturns/Point.hxx"
+#include "openturns/ResourceMap.hxx"
+#include "openturns/DistFunc.hxx"
+
+
+BEGIN_NAMESPACE_OPENTURNS
+
+CLASSNAMEINIT(DiscreteMarkovChain)
+
+static const Factory<DiscreteMarkovChain> Factory_DiscreteMarkovChain;
+
+/* Default constructor */
+DiscreteMarkovChain::DiscreteMarkovChain()
+  : ProcessImplementation()
+  , origin_()
+  , transitionMatrix_(IdentityMatrix(1))
+  , currentState_()
+{
+  // Set the dimension of the process
+  setOutputDimension(1);
+  
+  // Set the origin of the process
+  Dirac distribution(0);
+  setOrigin(distribution);
+  
+  // Set the current state of the process
+  currentState_ = 0;
+}
+
+/* Standard constructors */
+DiscreteMarkovChain::DiscreteMarkovChain(const Distribution & origin,
+                       const SquareMatrix & transitionMatrix)
+  : ProcessImplementation()
+  , origin_()
+  , transitionMatrix_()
+  , currentState_()
+{
+  // Set the dimension of the process
+  setOutputDimension(1);
+  
+  // Set the transition matrix of the process
+  setTransitionMatrix(transitionMatrix);
+  
+  // Set the origin of the process
+  setOrigin(origin);
+  
+  // Set the currentState of the process
+  currentState_ = origin.getRealization()[0];
+}
+
+DiscreteMarkovChain::DiscreteMarkovChain(const Distribution & origin,
+                       const SquareMatrix & transitionMatrix,
+                       const RegularGrid & timeGrid)
+  : ProcessImplementation()
+  , origin_()
+  , transitionMatrix_()
+  , currentState_()
+{
+  // Set the dimension of the process
+  setOutputDimension(1);
+  
+  // Set the transition matrix of the process
+  setTransitionMatrix(transitionMatrix);
+  
+  // Set the origin of the process
+  setOrigin(origin);
+  
+  // Set the currentState of the process
+  currentState_ = origin.getRealization()[0];
+
+  // Set the time grid of the process
+  setTimeGrid(timeGrid);
+}
+
+/* Constructors with fixed origin */
+DiscreteMarkovChain::DiscreteMarkovChain(const UnsignedInteger origin,
+                       const SquareMatrix & transitionMatrix)
+  : ProcessImplementation()
+  , origin_()
+  , transitionMatrix_()
+  , currentState_()
+{
+  // Set the dimension of the process
+  setOutputDimension(1);
+  
+  // Set the transition matrix of the process
+  setTransitionMatrix(transitionMatrix);
```

Comment:
If the transition matrix is set first, then it imposes its constraints to the origin distribution. By default, it is a Uniform() distribution, ie with range=[-1,1]. The lower bound is negative, it should throw an exception...